### PR TITLE
[fsdp] fix: sort buffers in fsdp2_load_full_state_dict to prevent NCCL deadlock with heterogeneous buffers

### LINKED ISSUE #6365: [bug][fsdp2] NCCL deadlock in fsdp2_load_full_state_dict with heterogeneous buffer sizes (Gemma4)
## System Info

- **veRL**: commit `e1ff774f` (main, May 15 2026)
- **Python**: 3.12.3
- **PyTorch**: 2.10.0+cu129
- **vLLM**: 0.19.1
- **Transformers**: 5.8.0
- **Hardware**: 8× NVIDIA H200 (143GB), Driver 580.95.05, CUDA 12.9
- **Model**: `google/gemma-4-26B-A4B-it` (text-only extraction, `Gemma4ForCausalLM`)

## Bug Description

`fsdp2_load_full_state_dict` deadlocks during the buffer broadcast phase when the model has buffers of **different sizes**. The function iterates `model.named_buffers()` and broadcasts each buffer from rank 0 to all other ranks. After FSDP2 wrapping, `named_buffers()` returns buffers in **non-deterministic order across ranks**. When buffers have different sizes, ranks end up calling `dist.broadcast` on mismatched tensors (e.g., rank 0 broadcasts a 256-element buffer while rank 1 expects a 128-element buffer on the same collective), causing an NCCL deadlock.

**Why Gemma4 but not other models:** Gemma4 uses dual attention (alternating sliding-window and global attention layers) with different rotary embedding dimensions — `full_attention_inv_freq` (256 elements) and `sliding_attention_inv_freq` (128 elements). This produces 35 buffers with heterogeneous sizes. Most other models (e.g., Qwen, Llama) have uniform buffer sizes, so even if ordering differs, the broadcast sizes still match.

## Root Cause

In `verl/utils/fsdp_utils.py`, `fsdp2_load_full_state_dict`:

```python
# rotary_emb is not in state_dict, so we need to broadcast it manually
for name, buf in model.named_buffers():
    dist.broadcast(buf, src=0)
```

After `fully_shard()`, `named_buffers()` iteration order is not guaranteed to be consistent across ranks. For Gemma4:

- Rank 0 might iterate: `layers.0.self_attn.full_attention_inv_freq` (256), `layers.0.self_attn.sliding_attention_inv_freq` (128), ...
- Rank 1 might iterate: `layers.0.self_attn.sliding_attention_inv_freq` (128), `layers.0.self_attn.full_attention_inv_freq` (256), ...

This means rank 0 broadcasts a 256-element tensor while rank 1 expects a 128-element tensor on the same NCCL collective → size mismatch → deadlock.

## Fix

Sort buffers by name before broadcasting to ensure deterministic order:

```python
# rotary_emb is not in state_dict, so we need to broadcast it manually
bufs = sorted(model.named_buffers(), key=lambda x: x[0])
for name, buf in bufs:
    dist.broadcast(buf, src=0)
```

## Evidence

| Scenario | Result |
|----------|--------|
| Gemma4 26B-A4B **without** sort fix | Hangs at init, NCCL timeout after `nccl_timeout` seconds |
| Gemma4 26B-A4B **with** sort fix | Passes init, reaches validation (but generates garbage) |
| Qwen3 8B **with** sort fix | Works normally, no regression |

The sort fix resolves the init deadlock, but Gemma4 still produces garbage output during validation (0% accuracy, dummy-weight behavior). This appears to be a separate issue with the weight transfer from FSDP to vLLM — `update_weights` completes successfully on all ranks, but the model generates as if running on uninitialized weights. We have confirmed that both FSDP2 and vLLM work correctly in isolation (standalone `load_weights` with real HF weights produces coherent output).

Also tested with Gemma4-31B (dense) and run into the same issues (and also the same garbage output in validation).

Came across this issue as well marked as closed https://github.com/verl-project/verl/issues/5999 which points to a separate issue with Gemma4 but in FSDP1 and I did not try it yet so not sure if it works there.

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

## Reproduction

```yaml
# verl/configs/grpo-gemma4-26b.yaml (minimal)
actor_rollout_ref:
  hybrid_engine: True
  model:
    path: google/gemma-4-26B-A4B-it  # or text-only extraction
    override_config:
      attn_implementation: sdpa
  actor:
    strategy: fsdp2
    fsdp_config:
      param_offload: false
  rollout:
    name: vllm
    gpu_memory_utilization: 0.4
    max_model_len: 20480
    enforce_eager: true
    tensor_model_parallel_size: 1

trainer:
  nnodes: 1
  n_gpus_per_node: 8
```

```bash
python -m verl.trainer.main_ppo --config-path=... --config-name=...
```

The run hangs indefinitely after `"After FSDP, memory allocated"` is logged, then times out with an NCCL BROADCAST error after `nccl_timeout` seconds.

With the sort() fix described above, we can bypass the timeout issue but the validation loop will produce garbage output outling a further issue downstream that may or may not be related. 

### Expected behavior

Here are the current behaviours of the Gemma4 runs with and without the parameter sorting fix with a comparison to the Qwen3 run that actually works.

### Gemma4 without sort fix — NCCL timeout

```
(WorkerDict pid=2416196) After FSDP, memory allocated (GB): 5.91, memory reserved (GB): 8.93, device memory used/total (GB): 10.41/139.80                 
(WorkerDict pid=2416196) Total steps: 340, num_warmup_steps: 0                                                                                            
(WorkerDict pid=2416196) [Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0                                       
(WorkerDict pid=2416196) [Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0                                       
(WorkerDict pid=2416196) [transformers] `Qwen2VLImageProcessorFast` is deprecated. The `Fast` suffix for image processors has been removed; use `Qwen2VLIm
ageProcessor` instead.    
[rank3]:[E515 18:04:22.241062569 ProcessGroupNCCL.cpp:688] [Rank 3] Watchdog caught collective operation timeout:
  WorkNCCL(SeqNum=690, OpType=BROADCAST, NumelIn=256, NumelOut=256, Timeout(ms)=1800000) ran for 1800046 milliseconds before timing out.
[rank3]:[E515 18:04:22.241391904 ProcessGroupNCCL.cpp:2277] [PG ID 0 PG GUID 0(default_pg) Rank 3]
  failure detected by watchdog at work sequence id: 690
[rank3]: Stack trace of the failed collective:
  #0 broadcast from torch/distributed/distributed_c10d.py:2907
  #1 wrapper from torch/distributed/c10d_logger.py:83
  #2 fsdp2_load_full_state_dict from verl/utils/fsdp_utils.py:492
  #3 _build_fsdp_module from verl/workers/engine/fsdp/transformer_impl.py:418
```

### Gemma4 with sort fix — passes init, reaches validation

```
Initial validation metrics:
  val-core/workplace_assistant/acc/mean@1: 0.0
  val-aux/num_turns/mean: 2.0
```

(0% accuracy is a separate weight-sync issue being investigated independently — the point is init completes without deadlock.)

### Qwen3 8B with sort fix — works normally

```
Initial validation metrics:
  val-core/workplace_assistant/acc/mean@1: 0.248
  val-aux/num_turns/mean: 7.824
```

### PR DESCRIPTION
**Pull request by nanastassacos**
Models with heterogeneous buffer sizes (e.g., Gemma4 with different rotary embedding dimensions for sliding vs full attention) cause NCCL deadlocks in fsdp2_load_full_state_dict because named_buffers() iteration order is non-deterministic across ranks after FSDP2 wrapping. 

Fix: sort buffers by name before broadcasting to ensure all ranks process them in the same order.

Also relaxes the FSDP wrap policy to skip layer classes not found in the model (e.g., vision/audio layers in text-only mode) instead of crashing.

Fixes #6365

### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl/pulls?q=is%3Apr+fsdp2_load_full_state_dict+buffer
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

Tested with Gemma4-26B-A4B (FSDP2, 8×H200). Without fix: hangs indefinitely at buffer broadcast. With fix: clean initialization and successful training. Logs displayed below

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

Fixes NCCL deadlock in `fsdp2_load_full_state_dict` for models with heterogeneous buffer sizes. 

After FSDP2 wrapping, `model.named_buffers()` iteration order is non-deterministic across ranks. For models with buffers of different sizes (e.g., Gemma4 has `full_attention_inv_freq`=256 and `sliding_attention_inv_freq`=128), different ranks broadcast mismatched tensors → NCCL deadlock.

**Fix:** Sort buffers by name before broadcasting to ensure deterministic order across all ranks. Also relaxes the FSDP wrap policy to skip layer classes not found in the model (e.g., multimodal models loaded in text-only mode) instead of crashing

- `verl/utils/fsdp_utils.py`: Sort `named_buffers()` by name before `dist.broadcast
- `verl/utils/fsdp_utils.py`: Warn instead of crash when configured layer classes are missing from model

### Logs
 
Without fix — hangs indefinitely:

[rank0] Broadcasting buffer model.layers.0.self_attn.rotary_emb.inv_freq (shape: [128])
[rank1] Broadcasting buffer model.layers.1.self_attn.rotary_emb.inv_freq (shape: [256])
NCCL deadlock: rank0 sends 128 elements, rank1 expects 256

With fix — clean initialization:

[rank0] Broadcasting buffer model.layers.0.self_attn.rotary_emb.inv_freq (shape: [128])
[rank1] Broadcasting buffer model.layers.0.self_attn.rotary_emb.inv_freq (shape: [128])
All ranks process buffers in same order

Affects: Gemma4, and any future model with heterogeneous buffer sizes under FSDP2

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs) - not required for this kind of change
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: Hard to unit test (requires multi-GPU NCCL). You can write: "Not feasible — requires multi-rank NCCL environment to reproduce the deadlock. Tested manually on 8×H200 with Gemma4-26B-A4B."
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by wuxibin89** [comment]
@nanastassacos Please resolve conflicts.

**Comment by nanastassacos** [comment]
@wuxibin89 Done - missed that a fix had already been merged.

**Review by wuxibin89 — DISMISSED** [review]

**Comment by wuxibin89** [comment]
@nanastassacos Please format code before submit PR: https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting

**Review by wuxibin89 — APPROVED** [review]
